File: tools/check_links.php
<?php
// Simple link checker for local files in the workspace.
// Usage: run `php tools/check_links.php` from repository root (c:\xampp\htdocs\SEB).

function findFiles($dir, $exts) {
    $rii = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($dir));
    $files = [];
    foreach ($rii as $file) {
        if ($file->isDir()) continue;
        $path = $file->getPathname();
        foreach ($exts as $ext) {
            if (strtolower(substr($path, -strlen($ext))) === strtolower($ext)) {
                $files[] = $path;
                break;
            }
        }
    }
    return $files;
}

function extractRefs($content) {
    $matches = [];
    $refs = [];
    // match src="..." and href="..."
    preg_match_all('/(?:src|href)\s*=\s*"([^"]+)"/i', $content, $matches);
    if (!empty($matches[1])) {
        foreach ($matches[1] as $m) $refs[] = $m;
    }
    return $refs;
}

$root = __DIR__ . DIRECTORY_SEPARATOR . '..';
$root = realpath($root);
if (!$root) {
    echo "Cannot determine project root\n"; exit(1);
}

$exts = ['.php','.html','.htm','.js','.css'];
$files = findFiles($root, $exts);

$missing = [];

foreach ($files as $file) {
    $content = @file_get_contents($file);
    if ($content === false) continue;
    $refs = extractRefs($content);
    if (!$refs) continue;
    $baseDir = dirname($file);
    foreach ($refs as $r) {
        // skip external and anchors and mailto and data URLs
        if (preg_match('#^(https?:)?//#i', $r)) continue;
        if (strpos($r, 'mailto:') === 0) continue;
        if (strpos($r, 'data:') === 0) continue;
        if (strpos($r, '#') === 0) continue;

        // remove query string and fragment
        $clean = preg_replace('/[\?#].*$/', '', $r);

        // absolute path (starting with /) -> relative to document root (assume project root)
        if (substr($clean,0,1) === '/') {
            $target = $root . str_replace('/', DIRECTORY_SEPARATOR, $clean);
        } else {
            $target = realpath($baseDir . DIRECTORY_SEPARATOR . $clean);
        }

        if ($target === false || !file_exists($target)) {
            $missing[] = [
                'source' => substr($file, strlen($root)+1),
                'ref' => $r,
                'resolved' => $target === false ? 'UNRESOLVED' : substr($target, strlen($root)+1)
            ];
        }
    }
}

if (empty($missing)) {
    echo "No missing local references found.\n";
    exit(0);
}

echo "Missing references found:\n\n";
foreach ($missing as $m) {
    echo "- Source: " . $m['source'] . "\n";
    echo "  Ref:    " . $m['ref'] . "\n";
    echo "  Resolved: " . $m['resolved'] . "\n\n";
}

echo "Total missing: " . count($missing) . "\n";

?>

File: tools/check_links.py
import os
import re
from pathlib import Path

def find_files(root_dir, extensions):
    """Find all files with given extensions"""
    files = []
    for ext in extensions:
        for fpath in Path(root_dir).rglob(f"*{ext}"):
            if fpath.is_file():
                files.append(str(fpath))
    return files

def extract_refs(content):
    """Extract src and href references from HTML/PHP/JS content"""
    matches = re.findall(r'(?:src|href)\s*=\s*"([^"]+)"', content, re.IGNORECASE)
    return matches

def main():
    root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    extensions = ['.php', '.html', '.htm', '.js', '.css']
    files = find_files(root, extensions)
    
    missing = []
    
    for fpath in files:
        try:
            with open(fpath, 'r', encoding='utf-8', errors='ignore') as f:
                content = f.read()
        except:
            continue
        
        refs = extract_refs(content)
        if not refs:
            continue
        
        base_dir = os.path.dirname(fpath)
        
        for ref in refs:
            # Skip external URLs, anchors, mailto, data URIs
            if ref.startswith('http') or ref.startswith('//'):
                continue
            if ref.startswith('mailto:') or ref.startswith('data:'):
                continue
            if ref.startswith('#'):
                continue
            
            # Remove query string and fragment
            clean = re.sub(r'[\?#].*$', '', ref)
            
            # Resolve target path
            if clean.startswith('/'):
                target = os.path.join(root, clean.lstrip('/'))
            else:
                target = os.path.normpath(os.path.join(base_dir, clean))
            
            if not os.path.exists(target):
                rel_source = os.path.relpath(fpath, root)
                rel_target = os.path.relpath(target, root) if os.path.exists(os.path.dirname(target)) else 'UNRESOLVED'
                missing.append({
                    'source': rel_source,
                    'ref': ref,
                    'resolved': rel_target
                })
    
    if not missing:
        print("✓ No missing local references found.")
        return
    
    print(f"✗ Found {len(missing)} missing references:\n")
    for m in missing:
        print(f"  Source:  {m['source']}")
        print(f"  Ref:     {m['ref']}")
        print(f"  Resolved: {m['resolved']}\n")
    
    print(f"Total missing: {len(missing)}")

if __name__ == '__main__':
    main()
